src/optimization/hooke_jeeves_method.cpp

```diff
@@ -0,0 +1,149 @@
+#include <k52/optimization/hooke_jeeves_method.h>
+
+#ifdef BUILD_WITH_MPI
+
+#include <boost/mpi.hpp>
+#include <k52/parallel/mpi/constants.h>
+
+#endif
+
+#include <iostream>
+#include <cmath>
+#include <algorithm>
+#include <stdexcept>
+
+#include <k52/optimization/params/i_continuous_parameters.h>
+
+using ::std::vector;
+
+namespace
+{
+	double kAcceleration = 3.5;
+}
+
+namespace k52
+{
+namespace optimization
+{
+
+HookeJeevesMethod::HookeJeevesMethod(double precision, size_t max_iteration_number,
+	double init_step)
+	: precision_(precision)
+	, max_iteration_number_(max_iteration_number)
+	, init_step_(init_step)
+{
+}
+
+HookeJeevesMethod* HookeJeevesMethod::Clone() const
+{
+	return new HookeJeevesMethod();
+}
+
+std::string HookeJeevesMethod::get_name() const
+{
+	return "Hooke-Jeeves Method";
+}
+
+#ifdef BUILD_WITH_MPI
+void HookeJeevesMethod::Send(boost::mpi::communicator* communicator, int target) const
+{
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, precision_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, init_step_);
+}
+
+void HookeJeevesMethod::Receive(boost::mpi::communicator* communicator, int source)
+{
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, precision_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, init_step_);
+}
+#endif
+
+vector<double> HookeJeevesMethod::FindOptimalParameters(const vector<double>& initial_parameters)
+{
+	vector<double> arguments(initial_parameters);
+	vector<double> arguments_after_search(initial_parameters.size());
+	vector<double> steps_array = vector<double>(arguments.size(), init_step_);
+	size_t iteration = 0;
+
+	do
+	{
+		arguments_after_search = CoordinatewiseSearch(arguments, steps_array);
+		if (arguments == arguments_after_search)
+		{
+			for (size_t i = 0; i < steps_array.size(); i++)
```

Comment:
It doesn't play a big role in modern c++ but prefix increment (++i) form is still preferable